src/java.desktop/unix/classes/sun/awt/X11/XCanvasPeer.java

```diff
@@ -62,36 +62,33 @@ public GraphicsConfiguration getAppropriateGraphicsConfiguration(
         if (graphicsConfig == null || gc == null) {
             return gc;
         }
-        // Opt: Only need to do if we're not using the default GC
 
-        int screenNum = ((X11GraphicsDevice)gc.getDevice()).getScreen();
+        final X11GraphicsDevice newDev = getSameScreenDevice(gc);
+        final int visualToLookFor = graphicsConfig.getVisual();
 
-        X11GraphicsConfig parentgc;
-        // save vis id of current gc
-        int visual = graphicsConfig.getVisual();
-
-        X11GraphicsDevice newDev = (X11GraphicsDevice) GraphicsEnvironment.
-            getLocalGraphicsEnvironment().
-            getScreenDevices()[screenNum];
-
-        for (int i = 0; i < newDev.getNumConfigs(screenNum); i++) {
-            if (visual == newDev.getConfigVisualId(i, screenNum)) {
-                // use that
-                graphicsConfig = (X11GraphicsConfig)newDev.getConfigurations()[i];
-                break;
+        final GraphicsConfiguration[] configurations = newDev.getConfigurations();
+        for (final GraphicsConfiguration config : configurations) {
+            final X11GraphicsConfig x11gc = (X11GraphicsConfig) config;
+            if (visualToLookFor == x11gc.getVisual()) {
+                graphicsConfig = x11gc;
             }
         }
-        // just in case...
-        if (graphicsConfig == null) {
-            graphicsConfig = (X11GraphicsConfig) GraphicsEnvironment.
-                getLocalGraphicsEnvironment().
-                getScreenDevices()[screenNum].
-                getDefaultConfiguration();
-        }
 
         return graphicsConfig;
     }
 
+    private X11GraphicsDevice getSameScreenDevice(GraphicsConfiguration gc) {
+        XToolkit.awtLock(); // so that the number of screens doesn't change during
+        try {
+            final int screenNum = ((X11GraphicsDevice) gc.getDevice()).getScreen();
+            return (X11GraphicsDevice) GraphicsEnvironment.
+                    getLocalGraphicsEnvironment().
+                    getScreenDevices()[screenNum];
+        } finally {
+            XToolkit.awtUnlock();
+        }
+    }
```

Comment:
I am not sure I understand the purpose of this method, it requests the current device for the GC, then requests the screen index of the device, and then request device by the screen index, what is the reason to do that? Why we cannot just use the device from the GC?

Reply:
This is just a named copy of lines 67 to 75. Since now this is the only piece of code that needs to work holding the AWT lock, I thought it best to isolate it.
As I understand, the purpose of the piece is to find the "new" `X11GraphicsDevice` with the screen number same to the given "old" `GraphicsConfiguration`.

Reply:
I don't understand why this was extracted out. It isn't called anywhere else and again makes the diff harder to understand.

Reply:
> I don't understand why this was extracted out. It isn't called anywhere else and again makes the diff harder to understand.

This AWT lock should be held for a very small portion of the original method. Initially, I left the code guarded by it in place and it looked weird and out of place. I understand that this is a judgement call, though.

As for the diffs, they are not very significant (essentially, 4 lines of code were re-grouped and moved around) and we only have to look at them here for a few days. The (refactored) code itself will be looked at by (presumably) a greater number of people over a far greater period of time. Less of a judgement call.

Reply:
>This is just a named copy of lines 67 to 75. Since now this is the only piece of code that needs to work holding the AWT lock, I thought it best to isolate it.
As I understand, the purpose of the piece is to find the "new" X11GraphicsDevice with the screen number same to the given "old" GraphicsConfiguration.

I am not sure that we can have some "old" and "new" devices which have the same screen id.
 * If list of devices was not changed then the "screenid" for old device is the index of that "old" device in the device array. So the new and old devices are the same.
 * If the list of devices was changed, then the "screenid" for the old device was invalidated, and it is again an index of that "old" device in the device array.

Please take a look to the fix for JDK-6804747, initially that method get the screen index as a parameter, and then replaced by the GC. The code of the method did not changed but the index just requested from the GC.

Probably the code is just buggies?

Reply:
@mrserb I agree with your logic, but fixing this now would certainly be not a semantically null change and would complicate the fix that has already been under review for over two months. I'd be happy to file a bug for this separate issue.

Reply:
The separate CR would be fine.

Reply:
@mrserb  I filed [JDK-8284645](https://bugs.openjdk.java.net/browse/JDK-8284645). Please, add more info there if necessary.